components/ws-manager-bridge/src/bridge.ts

```diff
@@ -504,23 +504,37 @@ export class WorkspaceManagerBridge implements Disposable {
             const actuallyRunningInstances = await client.getWorkspaces(ctx, new GetWorkspacesRequest());
             actuallyRunningInstances.getStatusList().forEach((s) => runningInstancesIdx.delete(s.getId()));
 
+            // runningInstancesIdx only contains instances that ws-manager is not aware of
             for (const [instanceId, ri] of runningInstancesIdx.entries()) {
                 const instance = ri.latestInstance;
-                if (instance.status.phase !== "running") {
-                    log.debug({ instanceId }, "Skipping instance", {
-                        phase: instance.status.phase,
-                        creationTime: instance.creationTime,
-                        region: instance.region,
-                    });
-                    continue;
+                const phase = instance.status.phase;
+
+                // When ws-manager is not aware of the following instances,
+                // they should be marked as stopped.
+                // If 'stopping', we give 10 sec estimated grace period in case of a race condition
+                if (
+                    phase === "pending" ||
+                    phase === "creating" ||
+                    phase === "initializing" ||
+                    phase === "running" ||
+                    (phase === "stopping" &&
+                        instance.stoppingTime &&
+                        durationLongerThanSeconds(Date.parse(instance.stoppingTime), 10))
+                ) {
+                    log.info(
+                        { instanceId, workspaceId: instance.workspaceId },
+                        "Database says the instance is running, but wsman does not know about it. Marking as stopped in database.",
+                        { installation },
+                    );
+                    await this.markWorkspaceInstanceAsStopped(ctx, ri, new Date());
```

Comment:
Definitely. I was thinking about how to test this better, because so far I tested it by cordoning the preview env node, but it feels lacking.

Logging first enable later could be a start. Could you elaborate on the separate branch idea?

Reply:
Also happy to have another pair of eyes when testing. So far my flow is:
1. Cordon the node
2. Run a workspace
3. See it stuck on pending
4. See stoppingTime and stoppedTime inserted to the DB ✅

But for some reason it takes a while until the phase is stopped.

Reply:
Maybe then you put in the old condition again and add the new if branch with just logging. We can review the log messages and then in another deploy activate the 'stopping'.

Reply:
YES let's do that 🚀